src/lsm/lsm_worker.c

```diff
@@ -93,11 +92,13 @@ __lsm_worker(void *arg)
 
 	cookie = (WT_LSM_WORKER_ARGS *)arg;
 	session = cookie->session;
-	conn = S2C(session);
 
 	entry = NULL;
-	while (F_ISSET(conn, WT_CONN_SERVER_RUN) &&
-	    F_ISSET(cookie, WT_LSM_WORKER_RUN)) {
+	for (;;) {
+		WT_BARRIER();			/* Don't cache the run flag. */
```

Comment:
What reason do you have to know these flags will be seen by the worker threads?

While I agree the existing code is likely to work, I don't believe that it's guaranteed to work, and a chunk of the changes in this branch fix potential races.